worker/src/socket.rs

```diff
@@ -0,0 +1,340 @@
+use std::task::Poll;
+
+use crate::r2::js_object;
+use crate::Result;
+use futures_util::FutureExt;
+use js_sys::{Boolean as JsBoolean, JsString, Number as JsNumber, Object as JsObject, Uint8Array};
+use wasm_bindgen::{JsCast, JsValue};
+use wasm_bindgen_futures::JsFuture;
+use worker_sys::console_log;
+
+enum Reading {
+    None,
+    Pending(JsFuture),
+    Ready(Vec<u8>),
+}
+
+impl Default for Reading {
+    fn default() -> Self {
+        Self::None
+    }
+}
+
+enum Writing {
+    Pending(JsFuture, usize),
+    None,
+}
+
+impl Default for Writing {
+    fn default() -> Self {
+        Self::None
+    }
+}
+
+enum Closing {
+    Pending(JsFuture),
+    None,
+}
+
+impl Default for Closing {
+    fn default() -> Self {
+        Self::None
+    }
+}
+
+/// Represents an outbound TCP connection from your Worker.
+pub struct Socket {
+    inner: worker_sys::Socket,
+    // TODO: Box this stuff so it can be sync / send?
+    _writable: worker_sys::WritableStream,
+    writer: worker_sys::WritableStreamDefaultWriter,
```

Comment:
I originally had it create a writer and reader on ever call to `poll_read`, which also seems wrong. With my current implementation, I think I could take the lock only when a read future is pending, and release after, but its not super clear to me what this buys you since everything is privately wrapped up in the `Socket` struct and I think most people will interact with it via `AsyncRead + AsyncWrite`.